subxt/src/blocks/blocks_client.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright 2019-2022 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+use crate::{
+    client::OnlineClientT,
+    error::Error,
+    rpc::Rpc,
+    Config,
+};
+use derivative::Derivative;
+use futures::{
+    future::Either,
+    stream,
+    Stream,
+    StreamExt,
+};
+use sp_runtime::traits::Header;
+use std::future::Future;
+
+/// A client for working with blocks.
+#[derive(Derivative)]
+#[derivative(Clone(bound = "Client: Clone"))]
+pub struct BlocksClient<T, Client> {
+    client: Client,
+    _marker: std::marker::PhantomData<T>,
+}
+
+impl<T, Client> BlocksClient<T, Client> {
+    /// Create a new [`BlocksClient`].
+    pub fn new(client: Client) -> Self {
+        Self {
+            client,
+            _marker: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<T, Client> BlocksClient<T, Client>
+where
+    T: Config,
+    Client: OnlineClientT<T>,
+{
+    /// Subscribe to blocks headers.
+    ///
+    /// # Note
+    ///
+    /// These blocks haven't necessarily been finalised yet. Prefer
+    /// [`BlocksClient::subscribe_finalized_headers()`] if that is important.
+    pub fn subscribe_headers(
+        &self,
+    ) -> impl Future<Output = Result<impl Stream<Item = Result<T::Header, Error>>, Error>>
+           + Send
+           + 'static {
+        let client = self.client.clone();
+        async move { client.rpc().subscribe_blocks().await }
+    }
+
+    /// Subscribe to finalized blocks headers and ensure that all missing blocks are filled.
+    ///
+    /// The Substrate's RPC does not guarantee that all finalized blocks are provided.
+    /// This function ensures that all missed blocks are fetched and returned in order.
+    ///
+    /// # Note
+    ///
+    /// The following is a possible scenario where Substrate does not generate an event
+    /// for the second block.
+    ///
+    /// ```sh
+    /// substrate      | Block A | --- | Block B | --- | Block C |
+    /// client         | Block A | ---             --- | Block C |
+    /// ```
+    ///
+    /// This function mitigates this by filling the gaps of missing blocks.
+    ///
+    /// ```sh
+    /// substrate      | Block A | --- | Block B | --- | Block C |
+    /// client         | Block A | --- | Block B | --- | Block C |
+    ///                                ^^^ filled
+    /// ```
```

Comment:
I wonder whether all of this detail is useful (it feels like an implementation detail that people would just "expect" anyway most of the time). I feel like it would be enough to say something like:

Subscribe to finalized block headers.

While the Substrate RPC method does not guarantee that all finalized block headers are provided, this function does.

Reply:
It makes sense, thanks for the suggestion!